chainstate/src/detail/mod.rs

```diff
@@ -281,101 +287,84 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
-    where
-        State: Clone,
-        MainAction: FnMut(
-            &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
-            &mut State,
-        ) -> Result<Res, BlockError>,
-        OnNewAttempt: FnMut(/*attempt_number:*/ usize),
-        OnDbErr: FnOnce(/*attempts_count:*/ usize, chainstate_storage::Error) -> BlockError,
-    {
-        // Note: the purpose of the Option here is just to get rid of extra clone at the beginning.
-        let mut state = None;
-
-        let result = self.with_rw_tx(
-            |chainstate_ref| main_action(chainstate_ref, state.insert(initial_state.clone())),
-            on_new_attempt,
-            on_db_err,
-        );
-
-        // Note: this "or_else" part is only possible in the degenerate case where with_rw_tx
-        // doesn't invoke the closure even once.
-        let state = state.unwrap_or_else(|| initial_state.clone());
-        (state, result)
-    }
-
-    /// Integrate the block into the blocktree, performing all the necessary checks and
-    /// updating `block_status` after each successful check.
+    /// Integrate the block into the blocktree, performing all the necessary checks.
     /// The returned bool indicates whether a reorg has occurred.
     fn integrate_block(
         chainstate_ref: &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
         block: &WithId<Block>,
-        block_index: &BlockIndex,
-        block_status: &mut BlockStatus,
-    ) -> Result<bool, BlockError> {
-        chainstate_ref.check_block_parent(block)?;
-        block_status.advance_validation_stage_to(BlockValidationStage::ParentOk);
-
-        chainstate_ref.check_block(block).map_err(BlockError::CheckBlockFailed)?;
-        block_status.advance_validation_stage_to(BlockValidationStage::CheckBlockOk);
-
-        // Note: we have to persist BlockIndex too, because it will be used
-        // by activate_best_chain below. There is no point in saving
-        // an intermediate BlockStatus though, so we ignore block_status here.
-        chainstate_ref.set_new_block_index(block_index)?;
-        chainstate_ref.persist_block(block)?;
+        block_index: BlockIndex,
+    ) -> Result<bool, BlockIntegrationError> {
+        let mut block_status = BlockStatus::new();
+
+        let result = chainstate_ref.check_block(block);
+        if result.is_err() {
+            // TODO: "technical" errors (e.g. a DB error) should not lead to permanent
+            // block invalidation. The same applies to the other unconditional
+            // call of "set_validation_failed" below.
+            // See https://github.com/mintlayer/mintlayer-core/issues/1033 (item #3).
+            block_status.set_validation_failed();
+        }
 
-        let best_block_id =
-            chainstate_ref.get_best_block_id().map_err(BlockError::BestBlockLoadError)?;
+        result
+            .map_err(BlockError::CheckBlockFailed)
+            .map_err(block_integration_error_constructor(block_status))?;
```

Comment:
imho this complicates code. Also it's not a constructor it's a specific constructor of OtherValidationError. I'd just construct an error here, more verbose but striaghtforward.